spec/controllers/admin/movies_controller_spec.rb

```diff
@@ -0,0 +1,87 @@
+require "rails_helper"
+
+RSpec.describe Admin::MoviesController, type: :controller do
+  context "when current user is not admin" do
+    before {get :index}
+
+    it "should redirect to login page" do
+      expect(response).to redirect_to new_user_session_path
+    end
+
+    it "should raise error" do
+      expect{raise CanCan::AccessDenied}.to raise_error(
+        "You are not authorized to access this page.")
+    end
+  end
+
+  context "when current user is admin" do
+    login_admin
+    before {get :index, format: :html}
+
+    it "should have a current_user" do
+      expect(subject.current_user).not_to be_nil
+    end
+
+    describe "GET #index" do
+      let(:movies){FactoryBot.create_list :movie, 2}
+
+      it "populates an array of movies" do
+        expect(assigns(:movies)).to match_array movies
+      end
+
+      it "should render the :index view" do
+        expect(response).to render_template :index
+      end
+    end
+
+    describe "ransack" do
+      let!(:movie1){FactoryBot.create :movie, {title_vi: "Ahihi Do Ngoc"}}
```

Comment:
chị dùng `let` có khác với `let!` ko em

Reply:
`let` chỉ khởi tạo duy nhất 1 lần và giá trị được khởi tạo đó sẽ được cache trong cùng it block. Biến sẽ không khởi tạo nếu biến không được gọi (lazy load).
`let!` sẽ được gọi trong before block, nghĩa là trước khi nhảy vào it block, giá trị của đối tượng sẽ được tạo sẵn.